File: native-ios/RideTrackerNative/CorePluginHost.swift
import Foundation

enum PluginCapability: String, CaseIterable, Codable, Hashable {
    case motionAcceleration = "motion.acceleration"
    case motionGyroscope = "motion.gyroscope"
    case motionOrientation = "motion.orientation"
    case locationPosition = "location.position"
    case locationSpeed = "location.speed"
    case locationAltitude = "location.altitude"
    case heartRateBpm = "heart-rate.bpm"
    case cameraPreview = "camera.preview"
    case cameraRecording = "camera.recording"
    case audioMicrophone = "audio.microphone"
    case videoImport = "video.import"
    case videoExport = "video.export"
    case telemetryImport = "telemetry.import"
    case telemetryExport = "telemetry.export"
    case calibrationMotion = "calibration.motion"
    case calibrationLocation = "calibration.location"
    case calibrationCamera = "calibration.camera"
    case hudWidgetSource = "hud.widget-source"
}

struct CorePluginDefinition: Identifiable, Equatable {
    let id: String
    let name: String
    let version: String
    let capabilities: Set<PluginCapability>

    init(id: String, name: String, version: String = "1.0.0", capabilities: Set<PluginCapability>) {
        self.id = id
        self.name = name
        self.version = version
        self.capabilities = capabilities
    }
}

final class CorePluginHost {
    private var plugins: [String: CorePluginDefinition] = [:]

    @discardableResult
    func register(_ plugin: CorePluginDefinition) -> CorePluginDefinition {
        precondition(!plugin.id.isEmpty, "plugin.id is required")
        precondition(!plugin.capabilities.isEmpty, "plugin.capabilities are required")
        precondition(plugins[plugin.id] == nil, "plugin already registered: \(plugin.id)")
        plugins[plugin.id] = plugin
        return plugin
    }

    func plugin(id: String) -> CorePluginDefinition? { plugins[id] }

    func list(capability: PluginCapability? = nil) -> [CorePluginDefinition] {
        plugins.values
            .filter { capability == nil || $0.capabilities.contains(capability!) }
            .sorted { $0.id < $1.id }
    }

    func clear() { plugins.removeAll() }

    func registerBuiltins() {
        guard plugins.isEmpty else { return }
        register(CorePluginDefinition(
            id: "internal-sensors",
            name: "Interne Smartphone-Sensoren",
            capabilities: [
                .motionAcceleration, .motionGyroscope, .motionOrientation,
                .locationPosition, .locationSpeed, .locationAltitude,
                .calibrationMotion, .calibrationLocation, .hudWidgetSource,
            ]
        ))
        register(CorePluginDefinition(
            id: "ble-heart-rate",
            name: "Bluetooth Herzfrequenz",
            capabilities: [.heartRateBpm, .hudWidgetSource]
        ))
        register(CorePluginDefinition(
            id: "external-imu",
            name: "Externe IMU",
            capabilities: [.motionAcceleration, .motionGyroscope, .motionOrientation, .calibrationMotion, .hudWidgetSource]
        ))
        register(CorePluginDefinition(
            id: "external-gnss",
            name: "Externer GNSS-Empfänger",
            capabilities: [.locationPosition, .locationSpeed, .locationAltitude, .calibrationLocation, .hudWidgetSource]
        ))
        register(CorePluginDefinition(
            id: "camera-source",
            name: "Kameraquelle",
            capabilities: [.cameraPreview, .cameraRecording, .audioMicrophone, .calibrationCamera]
        ))
    }
}

File: native-android/app/src/main/java/de/ridetracker/core/CorePluginHost.kt
package de.ridetracker.core

enum class PluginCapability(val id: String) {
    MOTION_ACCELERATION("motion.acceleration"),
    MOTION_GYROSCOPE("motion.gyroscope"),
    MOTION_ORIENTATION("motion.orientation"),
    LOCATION_POSITION("location.position"),
    LOCATION_SPEED("location.speed"),
    LOCATION_ALTITUDE("location.altitude"),
    HEART_RATE_BPM("heart-rate.bpm"),
    CAMERA_PREVIEW("camera.preview"),
    CAMERA_RECORDING("camera.recording"),
    AUDIO_MICROPHONE("audio.microphone"),
    VIDEO_IMPORT("video.import"),
    VIDEO_EXPORT("video.export"),
    TELEMETRY_IMPORT("telemetry.import"),
    TELEMETRY_EXPORT("telemetry.export"),
    DEVICE_DISCOVERY("device.discovery"),
    DEVICE_CONNECTION("device.connection"),
    CALIBRATION_MOTION("calibration.motion"),
    CALIBRATION_LOCATION("calibration.location"),
    CALIBRATION_CAMERA("calibration.camera"),
    HUD_WIDGET_SOURCE("hud.widget-source"),
}

data class CorePluginDefinition(
    val id: String,
    val name: String,
    val version: String = "1.0.0",
    val capabilities: Set<PluginCapability>,
)

class CorePluginHost {
    private val plugins = linkedMapOf<String, CorePluginDefinition>()

    fun register(plugin: CorePluginDefinition): CorePluginDefinition {
        require(plugin.id.isNotBlank()) { "plugin.id is required" }
        require(plugin.capabilities.isNotEmpty()) { "plugin.capabilities are required" }
        check(!plugins.containsKey(plugin.id)) { "plugin already registered: ${plugin.id}" }
        plugins[plugin.id] = plugin
        return plugin
    }

    fun get(id: String): CorePluginDefinition? = plugins[id]

    fun list(capability: PluginCapability? = null): List<CorePluginDefinition> = plugins.values
        .filter { capability == null || capability in it.capabilities }

    fun clear() = plugins.clear()

    fun registerBuiltins() {
        if (plugins.isNotEmpty()) return
        register(CorePluginDefinition(
            id = "internal-sensors",
            name = "Interne Smartphone-Sensoren",
            capabilities = setOf(
                PluginCapability.MOTION_ACCELERATION,
                PluginCapability.MOTION_GYROSCOPE,
                PluginCapability.MOTION_ORIENTATION,
                PluginCapability.LOCATION_POSITION,
                PluginCapability.LOCATION_SPEED,
                PluginCapability.LOCATION_ALTITUDE,
                PluginCapability.CALIBRATION_MOTION,
                PluginCapability.CALIBRATION_LOCATION,
                PluginCapability.HUD_WIDGET_SOURCE,
            ),
        ))
        register(CorePluginDefinition(
            id = "ble-heart-rate",
            name = "Bluetooth Herzfrequenz",
            capabilities = setOf(
                PluginCapability.HEART_RATE_BPM,
                PluginCapability.DEVICE_DISCOVERY,
                PluginCapability.DEVICE_CONNECTION,
                PluginCapability.HUD_WIDGET_SOURCE,
            ),
        ))
        register(CorePluginDefinition(
            id = "external-imu",
            name = "Externe IMU",
            capabilities = setOf(
                PluginCapability.MOTION_ACCELERATION,
                PluginCapability.MOTION_GYROSCOPE,
                PluginCapability.MOTION_ORIENTATION,
                PluginCapability.DEVICE_DISCOVERY,
                PluginCapability.DEVICE_CONNECTION,
                PluginCapability.CALIBRATION_MOTION,
                PluginCapability.HUD_WIDGET_SOURCE,
            ),
        ))
        register(CorePluginDefinition(
            id = "external-gnss",
            name = "Externer GNSS-Empfänger",
            capabilities = setOf(
                PluginCapability.LOCATION_POSITION,
                PluginCapability.LOCATION_SPEED,
                PluginCapability.LOCATION_ALTITUDE,
                PluginCapability.DEVICE_DISCOVERY,
                PluginCapability.DEVICE_CONNECTION,
                PluginCapability.CALIBRATION_LOCATION,
                PluginCapability.HUD_WIDGET_SOURCE,
            ),
        ))
        register(CorePluginDefinition(
            id = "camera-source",
            name = "Kameraquelle",
            capabilities = setOf(
                PluginCapability.CAMERA_PREVIEW,
                PluginCapability.CAMERA_RECORDING,
                PluginCapability.AUDIO_MICROPHONE,
                PluginCapability.CALIBRATION_CAMERA,
            ),
        ))
    }
}
